### ISSUE TITLE
undefined reference to vtable and typeinfo error

### ISSUE THREAD

**Issue by liushuya7** [jhu-bigss/maxon_epos2_driver#2]
I'm trying to compile **src/main_maxon_ui_control.cpp** using `colcon build` under a ROS2 workspace, but I got the following error:

```bash
--- stderr: maxon_epos2_driver                             
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonWidget.cpp.o: in function `cmnClassServicesBase* cmnClassServicesInstantiate<maxonWidget>()':
maxonWidget.cpp:(.text+0x162): undefined reference to `typeinfo for maxonWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonWidget.cpp.o: in function `maxonWidget::maxonWidget(maxonInterface*, QWidget*)':
maxonWidget.cpp:(.text+0x2e4): undefined reference to `vtable for maxonWidget'
/usr/bin/ld: maxonWidget.cpp:(.text+0x2f9): undefined reference to `vtable for maxonWidget'
/usr/bin/ld: maxonWidget.cpp:(.text+0x312): undefined reference to `vtable for maxonWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonWidget.cpp.o: in function `maxonWidget::~maxonWidget()':
maxonWidget.cpp:(.text+0xccb): undefined reference to `vtable for maxonWidget'
/usr/bin/ld: maxonWidget.cpp:(.text+0xcdd): undefined reference to `vtable for maxonWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonWidget.cpp.o:maxonWidget.cpp:(.text+0xcf3): more undefined references to `vtable for maxonWidget' follow
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonControlWidget.cpp.o: in function `cmnClassServicesBase* cmnClassServicesInstantiate<maxonControlWidget>()':
maxonControlWidget.cpp:(.text+0x168): undefined reference to `typeinfo for maxonControlWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonControlWidget.cpp.o: in function `maxonControlWidget::maxonControlWidget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, QWidget*)':
maxonControlWidget.cpp:(.text+0x2fc): undefined reference to `vtable for maxonControlWidget'
/usr/bin/ld: maxonControlWidget.cpp:(.text+0x311): undefined reference to `vtable for maxonControlWidget'
/usr/bin/ld: maxonControlWidget.cpp:(.text+0x32a): undefined reference to `vtable for maxonControlWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonControlWidget.cpp.o: in function `maxonControlWidget::~maxonControlWidget()':
maxonControlWidget.cpp:(.text+0x1bad): undefined reference to `vtable for maxonControlWidget'
/usr/bin/ld: maxonControlWidget.cpp:(.text+0x1bbf): undefined reference to `vtable for maxonControlWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonControlWidget.cpp.o:maxonControlWidget.cpp:(.text+0x1bd5): more undefined references to `vtable for maxonControlWidget' follow
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonStatusWidget.cpp.o: in function `cmnClassServicesBase* cmnClassServicesInstantiate<maxonStatusWidget>()':
maxonStatusWidget.cpp:(.text+0x168): undefined reference to `typeinfo for maxonStatusWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonStatusWidget.cpp.o: in function `maxonStatusWidget::maxonStatusWidget(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, QWidget*)':
maxonStatusWidget.cpp:(.text+0x2e0): undefined reference to `vtable for maxonStatusWidget'
/usr/bin/ld: maxonStatusWidget.cpp:(.text+0x2f2): undefined reference to `vtable for maxonStatusWidget'
/usr/bin/ld: maxonStatusWidget.cpp:(.text+0x308): undefined reference to `vtable for maxonStatusWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonStatusWidget.cpp.o: in function `maxonStatusWidget::~maxonStatusWidget()':
maxonStatusWidget.cpp:(.text+0x417): undefined reference to `vtable for maxonStatusWidget'
/usr/bin/ld: maxonStatusWidget.cpp:(.text+0x429): undefined reference to `vtable for maxonStatusWidget'
/usr/bin/ld: CMakeFiles/maxon_epos2_driver_gui.dir/src/maxonStatusWidget.cpp.o:maxonStatusWidget.cpp:(.text+0x43f): more undefined references to `vtable for maxonStatusWidget' follow
collect2: error: ld returned 1 exit status
gmake[2]: *** [CMakeFiles/maxon_epos2_driver_gui.dir/build.make:311: maxon_epos2_driver_gui] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:193: CMakeFiles/maxon_epos2_driver_gui.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< maxon_epos2_driver [4.32s, exited with code 2]
```

**Comment by jmz3** [comment]
I played with the cmakelists.txt for a while and it seems like only the GUI executable is influenced. I tried to comment out the `maxon_epos2_driver_gui`, the compilation was completed with no errors, and the `maxon_epos2_driver` was built successfully. I also tried to include Qt as an extra dependency, which didn't make a difference. I assume the problem must lie somewhere related to how the cisst define the GUI components. 

As far as I know, vtable errors occur when the class is inherited but the virtual methods are not defined or declared in the scope of the maxonWidget class. I don't know if declaring all the virtual methods would help in our case.

**Comment by jmz3** [comment]
Thsi issue is fixed. It turns out that QT is to be blamed. Some hints are provided [here](https://stackoverflow.com/questions/1552069/undefined-reference-to-vtable-trying-to-compile-a-qt-project). It's a known issue with GCC and Qt. 

TL;DR
Just insert the following line at the beginning (sort of, not exactly the first line) of the CMakeLists.txt would do the magic
```cmake
set (CMAKE_AUTOMOC ON)
```
And make sure to explicitly add the header files to the `add_library` and `add_executable` block to make sure AUTOMOC can locate the headers successfully. I will update the CMakeLists.txt in the future commits.
